src/js/components/devices/device-details/devicetwin.js

```diff
@@ -0,0 +1,215 @@
+
+const NoDiffStatus = ({ updateTime }) => (
+  <div className="padding" style={diffStatusStyle}>
+    <CheckCircleOutlined className="green" />
+    <div>No difference between desired and reported configuration</div>
+    <LastSyncNote updateTime={updateTime} />
+  </div>
+);
+
+export const TwinSyncStatus = ({ diffCount, updateTime }) =>
+  !diffCount ? (
+    <NoDiffStatus updateTime={updateTime} />
+  ) : (
+    <div className="padding" style={diffStatusStyle}>
+      <CloudUpload />
+      <div>
+        <b>
+          Found {diffCount} {pluralize('difference', diffCount)}
+        </b>{' '}
+        between desired and reported configuration
+      </div>
+      <LastSyncNote updateTime={updateTime} />
+    </div>
+  );
+
+const editorProps = {
+  height: 500,
+  defaultLanguage: 'json',
+  language: 'json',
+  loading: <Loader show />,
+  options: {
+    autoClosingOvertype: 'auto',
+    codeLens: false,
+    contextmenu: false,
+    enableSplitViewResizing: false,
+    formatOnPaste: true,
+    lightbulb: { enabled: false },
+    minimap: { enabled: false },
+    lineNumbersMinChars: 3,
+    quickSuggestions: false,
+    renderOverviewRuler: false,
+    scrollBeyondLastLine: false,
+    readOnly: true
+  }
+};
+
+const externalProvider = EXTERNAL_PROVIDER.azure;
+const indentation = 4; // number of spaces, tab based indentation won't show in the editor, but be converted to 4 spaces
+
+export const DeviceTwin = ({ device, setDeviceTwin }) => {
+  const [configured, setConfigured] = useState('');
+  const [diffCount, setDiffCount] = useState(0);
+  const [isEditing, setIsEditing] = useState(false);
+  const [errorMessage, setErrorMessage] = useState('');
+  const [open, setOpen] = useState(false);
+  const [reported, setReported] = useState('');
+  const [updated, setUpdated] = useState('');
+  const [isSync, setIsSync] = useState(true);
+  const editorRef = useRef(null);
+
+  const { [externalProvider.provider]: deviceTwin = {} } = device.twinsByProvider ?? {};
+  const { desired: configuredTwin, reported: reportedTwin, updated_ts: updateTime = device.created_ts } = deviceTwin;
+
+  useEffect(() => {
+    const textContent = JSON.stringify(configuredTwin, undefined, indentation) ?? '';
+    setConfigured(textContent);
+    setUpdated(textContent);
+    setReported(JSON.stringify(reportedTwin, undefined, indentation) ?? '');
+    setIsEditing;
+  }, [open]);
+
+  useEffect(() => {
+    setIsSync(deepCompare(reported, configured));
+  }, [configured, reported]);
+
+  const handleEditorDidMount = (editor, monaco) => {
+    editorRef.current = { editor, monaco, modifiedEditor: editor };
+  };
+
+  const handleDiffEditorDidMount = (editor, monaco) => {
+    const modifiedEditor = editor.getModifiedEditor();
+    modifiedEditor.onDidChangeModelContent(() => setUpdated(modifiedEditor.getValue()));
+    editor.onDidUpdateDiff(onDidUpdateDiff);
+    editorRef.current = { editor, monaco, modifiedEditor };
+  };
+
+  const onDidUpdateDiff = () => {
```

Comment:
The diff compares all JSON properties including the metadata such as when the properties that holds the individual update timestamp. Can we exclude properties starting with `$` from the diff?

Reply:
Since the entire object can be user defined I don't think excluding them all is a good idea, we might easily remove some user defined ones other wise... we could exclude the azure defined ones explicitly, but that still might cause trouble once a user defined one gets overwritten.

Reply:
As far as I can tell it's only the `$version` and `$metadata` properties in the root object that's set by the IoT Hub backend (although it looks like it's not always correctly set). However, these are not documented anywhere so I don't know what the correct approach is here :man_shrugging:
In any case, it will always show at least two lines differ because these properties will (almost) always mismatch.